niceman/distributions/venv.py

```diff
@@ -0,0 +1,211 @@
+from .base import Package
+from .base import SpecObject
+from .base import TypedList
+
+lgr = logging.getLogger("niceman.distributions.venv")
+
+
+@attr.s
+class VenvPackage(Package):
+    name = attr.ib()
+    version = attr.ib()
+    origin_location = attr.ib(default=None)
+    files = attr.ib(default=attr.Factory(list))
+
+
+@attr.s
+class VenvEnvironment(SpecObject):
+    name = attr.ib()
+    path = attr.ib(default=None)
+    python_version = attr.ib(default=None)
+    packages = TypedList(VenvPackage)
+
+
+@attr.s
+class VenvDistribution(Distribution):
+    """Class to provide virtualenv-based "distributions".
+    """
+    path = attr.ib(default=None)
+    venv_version = attr.ib(default=None)
+    environments = TypedList(VenvEnvironment)
+
+    def initiate(self, _):
+        return
+
+    def install_packages(self, session=None):
+        raise NotImplementedError
+
+
+class VenvTracer(DistributionTracer):
+    """Distribution tracer for virtualenv.
+    """
+
+    def _init(self):
+        self._path_root = PathRoot(self._is_venv_directory)
+
+    def _get_packagefields_for_files(self, files):
+        raise NotImplementedError
+
+    def _create_package(self, **package_fields):
+        raise NotImplementedError
+
+    def _get_package_details(self, venv_path):
+        packages = {}
+        file_to_pkg = {}
+
+        pkgs, locs = map(list, zip(*self._pip_packages(venv_path)))
+        batch = execute_command_batch(self._session,
+                                      [venv_path + "/bin/pip", "show", "-f"],
+                                      pkgs)
+        entries = (stacked.split("---") for stacked, _, _ in batch)
+
+        for pkg, loc, entry in zip(pkgs, locs, itertools.chain(*entries)):
+            info = parse_pip_show(entry)
+            details = {"name": info["Name"],
+                       "version": info["Version"],
+                       "location": loc}
+            packages[pkg] = details
+            for path in info["Files"]:
+                full_path = os.path.normpath(
+                    os.path.join(info["Location"], path))
+                file_to_pkg[full_path] = pkg
+        return packages, file_to_pkg
+
+    def _is_venv_directory(self, path):
+        try:
+            self._session.execute_command("grep -q VIRTUAL_ENV "
+                                          "{}/bin/activate".format(path))
+        except Exception as exc:
+            lgr.debug("Did not detect virtualenv at the path %s: %s",
+                      path, exc_str(exc))
+            return False
+        return True
+
+    def _get_venv_path(self, path):
+        return self._path_root(path)
+
+    def identify_distributions(self, files):
+        unknown_files = set(files)
+        found_package_count = 0
+        total_file_count = len(unknown_files)
+
+        venv_paths = map(self._get_venv_path, files)
+        venv_paths = set(filter(None, venv_paths))
+
+        short_names = abbreviate(venv_paths)
```

Comment:
hm, a handy function! but I think it would be fragile for some kind of incremental operation (e.g. the multipass we are contemplating).  

Also -- is having a name for venv is really needed?  it is identified uniquely by the path and it should be sufficient. the same was done in VCS (repos have no names)

Reply:
> Also -- is having a name for venv is really needed?

Hmm, can think of a reason why it's necessary.

Does the same argument apply to the conda tracer?

https://github.com/ReproNim/niceman/blob/30ae824fe3696a3a92ee34e28bf65a772cff8b15/niceman/distributions/conda.py#L385-L396

Reply:
conda environment actually has a name, but I am not quite sure why it is not the one given by `conda env list` here:
```shell
$> ~/anaconda2/bin/conda env list  
# conda environments:
#
conda-smmap              /home/yoh/anaconda2/envs/conda-smmap
datalad                  /home/yoh/anaconda2/envs/datalad
heudiconv                /home/yoh/anaconda2/envs/heudiconv
pymvpa                   /home/yoh/anaconda2/envs/pymvpa
root                  *  /home/yoh/anaconda2
```
e.g. the main one should be `root'.

As for the name of the entire distribution (`CondaDistribution`), it is indeed probably to be RFed: #170